CMakeLists.txt

```diff
@@ -46,4 +46,85 @@ if(Redpanda_ENABLE_COVERAGE)
   add_link_options(--coverage)
 endif()
 
+#
+# Configure clang-tidy checks. By default checks will be enabled if the
+# clang-tidy tool is found. Otherwise, it can be explicitly disabled. It's
+# recommended to use FORCE_ON in contexts such as CI where we want to fail
+# configuration if clang-tidy is not available.
+#
+set(Redpanda_ENABLE_CLANG_TIDY "ON" CACHE STRING
+  "Enable clang-tidy checks if available. Can be ON, OFF, or FORCE_ON")
+
+if(NOT Redpanda_ENABLE_CLANG_TIDY STREQUAL OFF)
+  if(Redpanda_ENABLE_CLANG_TIDY STREQUAL FORCE_ON)
+    find_program(CLANG_TIDY_COMMAND clang-tidy
+      PATHS ${PROJECT_SOURCE_DIR}/vbuild/llvm/install/bin
+      REQUIRED)
+  else()
+    find_program(CLANG_TIDY_COMMAND clang-tidy
+      PATHS ${PROJECT_SOURCE_DIR}/vbuild/llvm/install/bin)
+  endif()
+  if(CLANG_TIDY_COMMAND)
+    set(Redpanda_ENABLE_CLANG_TIDY ON)
+  else()
+    set(Redpanda_ENABLE_CLANG_TIDY OFF)
+  endif()
+endif()
+
+# Default clang-tidy checks for use in non-test code.
+#
+# All checks are enabled, with the following exceptions:
+#
+#   * altera-*, fuschia-*, llvmlibc-*: project specific checks
+#   * llvm-header-guard: we use `#pragma once`
+#   * -modernize-use-trailing-return-type: arbitrary style choice
+#
+set(clang_tidy_default_checks
+  *
+  -llvm-header-guard
+  -modernize-use-trailing-return-type
+  -altera-*
+  -fuchsia-*
+  -llvmlibc-*)
+
+# Default clang-tidy checks for test code.
+#
+# All non-test default checks are enabled, with the following exceptions:
+#
+#   * readability-function-cognitive-complexity: triggers for TEST_F macro
+#   * readability-identifier-length: quality of life improvement
+#   * cppcoreguidelines-avoid-magic-numbers: qualifty of life improvement
+#   * readability-magic-numbers: duplicate
+#   * cppcoreguidelines-pointer-arithmetic: quality of life improvement
+#
+set(clang_tidy_test_checks
+  ${clang_tidy_default_checks}
+  -readability-function-cognitive-complexity
+  -readability-identifier-length
+  -cppcoreguidelines-avoid-magic-numbers
+  -readability-magic-numbers
+  -cppcoreguidelines-pro-bounds-pointer-arithmetic)
+
+if(Redpanda_ENABLE_CLANG_TIDY)
+  macro(enable_clang_tidy)
+    list(JOIN clang_tidy_default_checks "," clang_tidy_checks)
+    set(CMAKE_CXX_CLANG_TIDY ${CLANG_TIDY_COMMAND}
+      -warnings-as-errors=* -checks=${clang_tidy_checks})
+    unset(clang_tidy_checks)
+  endmacro()
+
+  macro(enable_clang_tidy_for_tests)
+    list(JOIN clang_tidy_test_checks "," clang_tidy_checks)
+    set(CMAKE_CXX_CLANG_TIDY ${CLANG_TIDY_COMMAND}
```

Comment:
Will clangd pick these up? ie. do they end up in the compilation database?

Reply:
not that i know of. that could be one reason to try to do all of this in the .clang-tidy file. i don't know if it is powerful enough let us do all the directoy inclusion/exclusion trickery, tho.

Reply:
so it seems like the way to do this is to add a .clang-tidy file into each sub-directory, and the one at the root will apply by default. so the root one stays the same, and then we can add the stricter checks into each sub-directory. that should be picked up by both clangd and this cmake thing. we can then symlink them to avoid lots of duplication.

it may be quite a long time before the whole tree is clang-tidy clean, so we may end up with lots of these files for some time, but thats not a big deal?

Reply:
We can symlink them, but I don't think it's a big deal.